File: go_code/server/init_app.go
package main

import (
	"log"
	"os"

	"server/interfaces"
	"server/services"
	"server/storage"

	"github.com/gin-gonic/gin"
)

// getStorageType returns the storage type from environment
func getStorageType() string {
	storageType := os.Getenv("STORAGE_TYPE")
	if storageType == "" {
		return "memory"
	}
	return storageType
}

// createStore creates a new store based on configuration
func createStore() interfaces.KeyValueStore {
	storageType := getStorageType()

	switch storageType {
	case "redis":
		redisAddr := os.Getenv("REDIS_ADDR")
		if redisAddr == "" {
			redisAddr = "localhost:6379"
		}
		log.Printf("Using Redis storage at %s", redisAddr)
		return storage.NewRedisStore(redisAddr)
	default:
		log.Println("Using in-memory storage")
		return storage.NewInMemoryStore()
	}
}

// getService creates a new service instance
func getService() interfaces.KeyValueService {
	store := createStore()
	return services.NewKeyValueService(store)
}

// serviceMiddleware injects the service into the Gin context
func serviceMiddleware() gin.HandlerFunc {
	return func(c *gin.Context) {
		service := getService()
		c.Set("service", service)
		c.Next()
	}
}

File: server/init_app.py
import os
import redis
from fastapi import Depends
from services.key_value_service import KeyValueService
from storage.key_value_store import KeyValueStore
from storage.redis_store import RedisStore
from storage.in_memory_store import InMemoryStore


"""
Get the storage type for startup (memory or redis)
"""
def get_storage_type() -> str:
    return os.getenv("STORAGE_TYPE", "memory")  


"""
Create the store based on the storage type
"""
def create_store() -> KeyValueStore:
    storage_type = get_storage_type()
    
    if storage_type == "redis":
        redis_client = redis.Redis(
            host=os.getenv("REDIS_HOST", "localhost"),
            port=int(os.getenv("REDIS_PORT", 6379)),
            db=int(os.getenv("REDIS_DB", 0)),
            decode_responses=True
        )
        return RedisStore(redis_client)
    else:
        return InMemoryStore()


"""
Global store instance
"""
store_instance = None


"""
Get the store
"""
def get_store() -> KeyValueStore:
    global store_instance
    if store_instance is None:
        store_instance = create_store()
    return store_instance


"""
Get the service
"""
def get_service(store: KeyValueStore = Depends(get_store)) -> KeyValueService:
    return KeyValueService(store)